src/codegen/mod.rs

```diff
@@ -563,131 +770,209 @@ fn gen_par_for(
         par_section
     };
 
-    cu::Stmt::Seq(Box::new(body), Box::new(sync_stmt))
+    cu::Stmt::Seq (
+        Box::new(body),
+        Box::new(sync_stmt),
+    )
+}
+
+fn gen_checked_stmt(
+    expr: &desc::Expr,
+    view_ctx: &mut HashMap<String, ViewExpr>,
+    comp_unit: &[desc::FunDef],
+    label: Option<String>,
+) -> cu::Stmt {
+    use desc::ExprKind::*;
+
+    match &expr.expr {
+        Let(_, _, _, expr1, _) => gen_checked_stmt(expr, view_ctx, comp_unit, label), //TODO call on other expr of let
+        Seq(e1, e2) => {
+            cu::Stmt::Seq(
+                Box::new(gen_checked_stmt(e1, view_ctx, comp_unit, label.clone())),
+                Box::new(gen_checked_stmt(e2, view_ctx, comp_unit, label))
+            )
+        }
+        Index(_, i) => {
+            match label {
+                Some(l) => {
+                    // TODO implement arr size
+                    // let n = match pl_expr.ty.unwrap() {
+                    //     Ty::DataTy(DataTy::Array(_ , m)) => m,
+                    //     Ty::ViewTy(ViewTy::Array(_, m)) => m,
+                    //     _ => panic!("cannot index into non array type!"),
+                    // }
+                    use crate::ast::*;
+                    cu::Stmt::IndexCheck {
+                        // arr: Box::new(gen_pl_expr(pl_expr, view_ctx, comp_unit)),
+                        size: Nat::Lit(100),
```

Comment:
at the moment all check whether the index is smaller than 100. I will implement L.798-803 in the coming days